syntax = "proto3";

package hourai.data;

import "hourai/data/models.proto";
import "hourai/data/actions.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";

// The per-guild level configuration for the bot's administrative
// functions
message AdminConfig {
  EscalationLadder ladder = 1;
  uint64 moderation_role_id = 2;
  uint64 log_channel_id = 3;
  repeated uint64 self_serve_role_ids = 4;
}

// The escalation ladder for a server
message EscalationLadder {
  // The list of rungs on the ladder
  repeated EscalationRung rung = 1;
}

// A rung on the escalation ladder
message EscalationRung {
  // The moderation action that is to be applied at this level
  hourai.actions.Action action = 1;
  // The time it takes to deescalate from this run on the ladder,
  // in seconds
  google.protobuf.Duration deescalation_period = 2;
}

// The current state of a user subject to previous moderation
message ModeratedUser {
  hourai.models.MemberId member_id = 1;
  // The last few mod actions taken regarding the user
  repeated ModerationEntry last_mod_action = 2;
  // The current escalation level of the user.
  int32 current_level = 3;
  // The timestamp when the user will next be deescalated,
  // if not present, the user will not be deescalated further.
  google.protobuf.Timestamp next_deescalation = 4;
}

// An moderation audit log entry
message ModerationEntry {
  hourai.models.MemberId member_id = 1;
  // The action that was applied to the user
  hourai.actions.Action action = 2;
  // The escalation level the user reached as a result of this
  // action
  int64 level = 6;
}

